controllers/secret_controller.go

```diff
@@ -80,6 +81,9 @@ func (s *SecretSyncController) Reconcile(ctx context.Context, req ctrl.Request)
 
 				secretsToDelete := make([]*corev1.Secret, 0, len(secrets))
 
+				s.selectorMutex.Lock()
```

Comment:
This makes me nervous. We are holding the lock the entire time we iterate over secrets. While the number of secrets is probably going to be small in number, the actual number is arbitrary. We also perform deletes against the remote clusters in this loop. We have no idea what latencies could be like with those remote deletes. Any goroutine that tries to update the cache in the `toMatchingReplicates` function will be blocked. This may impact other controllers as well. What about making a copy of the cache here and using that in the loop?

Reply:
I can modify this to use RWLock(), that shouldn't prevent adding new ones to the queue.

Reply:
Are you referring to a RWMutex? If so, isn't it still a mutually exclusive lock with respect to readers and writers?

Reply:
Yes, but we don't hold read lock for a long time (and don't do any calls to api-server while we hold it). 

Multiple readers can have the read lock at the same time, so they don't block each other.

Reply:
I realize that we don't hold the write lock for long and also that we can have multiple readers. My concern is that the loop could have the read lock for some time and block writes in `toMatchingReplicates`. Switching to a lock that allows multiple readers doesn't address this.

Reply:
toMatchingReplicas does not use write lock

Reply:
ah, I had to update my branch. I see the change. 👍